presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -375,28 +431,39 @@ public FragmentProperties setCoordinatorOnlyDistribution()
             return this;
         }
 
-        public FragmentProperties addSourceDistribution(PlanNodeId source, PartitioningHandle distribution)
+        public FragmentProperties addSourceDistribution(PlanNodeId source, PartitioningHandle distribution, Metadata metadata, Session session)
```

Comment:
Now two methods will `coalesce` fragment's distribution:
- `setDistribution`
- `addSourceDistribution`

Maybe consider to have some common method to do `coalesce`. This might help reason the code ? (e.g. when is fragment coalesce happen)